### ISSUE TITLE
Legion: seg fault in dependence analysis

### ISSUE THREAD

**Issue by MadFunMaker** [StanfordLegion/legion#1452]
## Environment

I ran FlexFlow with the following Legion version on the multi-node environment.

branch: control_replication
commit ID: 3354a70192188e5eca7d0c4f96fc3b370541c8fc

Here is the experiment configuration for the example above:
- Summit cluster
- Number of nodes / GPUs: 4 / 16 (4 GPUs per node)
- Number of tasks per GPU: 64
- Total number of tasks: 1024 (64 tasks / GPU * 16 GPUs).

------------------------------------------------------------------------------------------------------------------------
## Issues

May I get some advice on 1) why this segfault would happen and 2) how to investigate it? 

I got a segmentation fault in the following function:
`Legion::Internal::Operation::perform_registration(unsigned int, Legion::Internal::Operation*, unsigned int, bool&, Legion::Internal::Operation::MappingDependenceTracker*, Legion::Internal::RtEvent) ()`

For further detail, here is the full backtrace of the process/thread (that calls nanosleep()) from gdb: 

**Process backtrace**
```
#0  0x0000200016655e84 in syscall () from /lib64/power9/libc.so.6
#1  0x000020001456e2a4 in Realm::Doorbell::wait_slow() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#2  0x000020001456fa08 in Realm::UnfairCondVar::wait() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#3  0x00002000144a4cb8 in Realm::RuntimeImpl::wait_for_shutdown() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#4  0x00002000144a675c in Realm::Runtime::wait_for_shutdown() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#5  0x0000200012e632a8 in Legion::Internal::Runtime::start(int, char**, bool, bool) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#6  0x0000200012a45ae8 in Legion::Runtime::start(int, char**, bool, bool) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#7  0x0000000010008078 in main ()
```

**Thread backtrace**

```
#0  0x000020001661a114 in nanosleep () from /lib64/power9/libc.so.6
#1  0x0000200016619f44 in sleep () from /lib64/power9/libc.so.6
#2  0x00002000144a12a8 in Realm::realm_freeze(int) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#3  <signal handler called>
#4  0x00002000128b8a6c in std::_Rb_tree<Legion::Internal::RtEvent, Legion::Internal::RtEvent, std::_Identity<Legion::Internal::RtEvent>, std::less<Legion::Internal::RtEvent>, std::allocator<Legion::Internal::RtEvent> >::_M_get_insert_unique_pos(
Legion::Internal::RtEvent const&) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#5  0x00002000128b8b78 in std::pair<std::_Rb_tree_iterator<Legion::Internal::RtEvent>, bool> std::_Rb_tree<Legion::Internal::RtEvent, Legion::Internal::RtEvent, std::_Identity<Legion::Internal::RtEvent>, std::less<Legion::Internal::RtEvent>, std
::allocator<Legion::Internal::RtEvent> >::_M_insert_unique<Legion::Internal::RtEvent const&>(Legion::Internal::RtEvent const&) () from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#6  0x0000200012ae42c4 in Legion::Internal::Operation::perform_registration(unsigned int, Legion::Internal::Operation*, unsigned int, bool&, Legion::Internal::Operation::MappingDependenceTracker*, Legion::Internal::RtEvent) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#7  0x0000200012ae724c in Legion::Internal::Operation::register_dependence(Legion::Internal::Operation*, unsigned int) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#8  0x0000200012954090 in Legion::Internal::InnerContext::register_implicit_dependences(Legion::Internal::Operation*) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#9  0x0000200012a97e2c in Legion::Internal::Operation::begin_dependence_analysis() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#10 0x0000200012ad0cd4 in Legion::Internal::Operation::execute_dependence_analysis() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#11 0x000020001299aafc in Legion::Internal::InnerContext::process_dependence_stage() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#12 0x000020001299ad0c in Legion::Internal::InnerContext::handle_dependence_stage(void const*) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#13 0x0000200012e6d7d8 in Legion::Internal::Runtime::legion_runtime_task(void const*, unsigned long, void const*, unsigned long, Realm::Processor) () from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/liblegion.so.1
#14 0x000020001447cf30 in Realm::LocalTaskProcessor::execute_task(unsigned int, Realm::ByteArrayRef const&) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#15 0x00002000144d9fc0 in Realm::Task::execute_on_processor(Realm::Processor) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#16 0x00002000144da184 in Realm::UserThreadTaskScheduler::execute_task(Realm::Task*) ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#17 0x00002000144dd174 in Realm::ThreadedTaskScheduler::scheduler_loop() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#18 0x00002000144e6790 in Realm::UserThread::uthread_entry() ()
   from /ccs/home/jbsimdicd/gpp/ff-gpp-legion-debug/build/deps/legion/lib/librealm.so.1
#19 0x0000200016587ffc in makecontext () from /lib64/power9/libc.so.6
#20 0x0000000000001c3e in ?? ()
```

FYI, I had segfault only in multi-node configuration so far; specifically, it deterministically failed when # of nodes = 4, 8 while it worked for # of nodes = 1, 2. Another thing to note for our application is that we have a lot of tasks (64 tasks / GPU, 1024 tasks in total). 

Feel free to let me know if you need more info or context.

**Comment by lightsighter** [comment]
This is a memory corruption. I need a reproducer to look at it, but in general memory corruptions are always the application's fault.

**Comment by MadFunMaker** [comment]
@lightsighter Thank you for letting us know. Then, I will check the application code first.

**Comment by lightsighter** [comment]
Other than turning on bounds checks (only valid if you aren't getting raw pointers from your accessors), there's not much you're going to be able to do to look for memory corruption in a multi-node setting since you can't run valgrind. I suggest trying to make a minimal reproducer.

**Comment by MadFunMaker** [comment]
@lightsighter Per your advice, I generated a reproducer on the Sapling server. Unfortunately, it requires 4 nodes and ~ 30mins to get SegFault since I couldn't get the smaller reproducer so far. Another thing to note is that I built with "RelWithDebInfo"(not "Debug") mode since the same program had no SegFault in Debug mode. We also found that SegFault didn't happen for 4 nodes with a lighter workload if helpful. It always happened for the cases of 8 nodes.

You can do the following to reproduce it with RelWithDebInfo mode:
```
bash /home/shicao/eval/scripts/run_exp_n4_segfault.sh
```

FYI, Legion is located in "/home/shicao/ff-gpp/deps/legion". If you need anything else from me, feel free to let me know.

**Comment by MadFunMaker** [comment]
We observe Segfault happens when we replay dependencies that were recorded by tracing. Interestingly, if we replay a single iteration, the program ends with no Segfault. However, if we replay 10 iterations, then SegFault (that I described above) arises.

**Comment by lightsighter** [comment]
I ran this twice last night and didn't fail either time. Please run it for me and send me the process IDs for a frozen process on sapling when it fails.

**Comment by MadFunMaker** [comment]
@lightsighter Thank you for your help. I will send you that info as soon as GPU is available on Sapling. Would it be ok to run with Release mode only if I also fail to get SegFault with RelWithDebInfo mode?

**Comment by lightsighter** [comment]
No, it has to have debug symbols in the binary for me to be able to debug it, which requires at least `RelWithDebInfo`.

**Comment by MadFunMaker** [comment]
Ok. I will make sure to include debug symbols.

**Comment by MadFunMaker** [comment]
@lightsighter Here are the process ID and node: 
```
ssh g0004
gdb -p 3829754
```

For your information, here is the error message:
```
[3 - 7f3240046000] 2287.299516 {5}{runtime}: [error 39] LEGION ERROR: Invalid color space color for child 139776153878688 of partition 1017 (from file /home/shicao/ff-gpp-debug-shiyi/deps/legion/runtime/legion/region_tree.cc:9632)
For more information see:
http://legion.stanford.edu/messages/error_code.html#error_code_39

Legion process received signal 6: Aborted
Process 3829754 on node g0004.stanford.edu is frozen!
```
FYI, I didn't see the "invalid color space" error message on the Summit cluster if I am not wrong. And I was able to reproduce this issue twice in a row with the following command:
`bash /home/shicao/eval/scripts/run_exp_n4_segfault.sh`

The change was to take out `CXXFLAGS=-DLEGION_SPY`; weirdly, I noticed that SegFault didn't arise on sapling if I add this.

**Comment by lightsighter** [comment]
More heap memory corruption. The STL vector containing the path for walking through the region tree was corrupted. Whether you use Legion Spy or not is a red-herring. It's just perturbing the timing which changes whether you observe the memory corruption or not.  You can kill the job, we're not going to learn anything interesting from it. 

Try commenting out all the bodies of your tasks (and I do mean all of them) so that they are empty so we are just running Legion code with empty tasks. If you can get that to corrupt the heap then the memory corruption is the runtime's fault. If we don't see any corruption from repeated runs while running in that mode, then it is almost certainly the application's fault for corrupting the heap.

**Comment by MadFunMaker** [comment]
@lightsighter Thank you for your comment. I will try that. FYI, I have never seen Segfault though if I comment out the part replaying traces although this might just be perturbing the timing as well.

**Comment by MadFunMaker** [comment]
We checked that SegFault doesn't arise if we comment out all the bodies of our tasks, which means it's almost certainly the application's fault, as you said. We suspect it has something to do with the accessor on the application code. We will look into it by ourselves first and let you know.

**Comment by lightsighter** [comment]
If you think it is an out-of-bounds access and you're using Legion `FieldAccessors` for all your accesses then you can turn on bounds checks. It will slow things down but will report any out-of-bounds accesses. If you're getting raw pointers from your accessors then the analysis is no longer sound.

FWIW, I suspect that an out-of-bounds access from an accessor is not very likely to cause heap corruption (unless you've attached lots of external memory allocations to logical regions) because the part of the heap where instances are is usually isolated by a lot of distance from the rest of the heap (Realm pre-allocates all the space in its memories). Heap corruption of STL data structures more likely comes from something use doing something bad with an STL (or STL-like) data structure somewhere else in the program.

**Comment by MadFunMaker** [comment]
@lightsighter We found that running Legion code with empty tasks still causes non-deterministic Segfault, which indicates it's a runtime's fault, according to your comment. However, I couldn't create a reproducer in Sapling so far (which makes me believe SegFault doesn't arise with empty tasks last time when I said that to you. Sorry for the confusion). Could you share some tips to debug this issue in case I can't get a reproducer on Sapling? FYI, here is some additional information about this issue I got over the last two weeks. For clarity, let N be the number of nodes (4 GPUs per node).

- SegFault always arises if N=8 while it comes and goes if N=4; it never happens if N<4 even when I increase the number of tasks even more than the number of tasks in N=8; I wanted to reproduce it in a single node to use Valgrind, but I couldn't.
- We couldn't find any out-of-bound access from the application code.
- SegFault always arises when we replay the recorded trace.

**Comment by MadFunMaker** [comment]
I think I can make a reproducer with empty tasks on Sapling by tonight or tmr. I will let you know.

**Comment by MadFunMaker** [comment]
@lightsighter We found that running Legion with empty tasks still causes non-deterministic Segfault, which indicates it's a runtime's fault, according to your comment. I created a minimal reproducer with empty tasks on 2 GPU nodes of Sapling (RelWithDebInfo mode). You can use the following command if you want to reproduce it by yourself:
`bash /home/shicao/eval/scripts/run-exp-n2-segfault-no-fusion.sh`

FYI, here is some additional information about this issue I got over the last two weeks.
- We couldn't find any out-of-bound access from the application code.
- SegFault seems to arise only when we replay the recorded trace.

**Comment by lightsighter** [comment]
It's something in the transitive reduction of the trace that is stomping on the heap (which is unfortunately code that I did not write). The data structure that got stomped on very clearly has markers from the transitive reduction in it:

```
path = std::vector of length 4, capacity 4 = {0x0, 0x0, 0x0, 0x7fffffffffffffff},
```

The things in that path are supposed to be colors in a region tree, but `0x7fffffffffffffff` is `-1U` which is used as a guard for parts of the transitive reduction. And we can see the transitive reduction running in parallel with this code:

```
(gdb) bt
#0  Legion::Internal::PhysicalTemplate::transitive_reduction (this=0x7fc354ba37a0, deferred=<optimized out>) at /usr/include/c++/9/bits/stl_algobase.h:222
#1  0x00007fd58ec0523c in Legion::Internal::Runtime::legion_runtime_task (args=0x7fc2801958d0, arglen=12, userdata=<optimized out>, userlen=<optimized out>, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.cc:32461
```

Unfortunately with the way that you've compiled it we can't see the line numbers with the part of the algorithm that we are in. @MadFunMaker can you rebuild in debug mode with no optimizations on so we can see which phase of the transitive reduction we are in?

This is going to be really painful to debug because this trace is monstrously large. Note this is such a huge trace that the transitive reduction is probably never going to finish before your code is done running.

```
(gdb) p instructions.size()
$88203 = 187832
(gdb) p incoming.size()
$88204 = 427314
(gdb) p outgoing.size()
$88205 = 427314
```

@magnatelee Do you have any better suggestions than just putting bounds checks on all the data structure accesses in the transitive reduction?

**Comment by MadFunMaker** [comment]
@lightsighter Thank you very much for your detailed comment. I rebuilt in debug mode with no optimizations (i.e., GCC -O0) as you requested. Plus, I was able to reduce the size of the reproducer (= size of trace) further in debug mode; the minimal one only requires a single node now. You can try the following command:
```
bash /home/shicao/eval/scripts/run-exp-n1-segfault-no-fusion.sh 64
```
I checked that this leads to SegFault twice out of 3 times. Note that SegFault is non-deterministic. From my experience, you can increase the chance of SegFault by increasing the first argument (loosely, the number of tasks) to a bigger power of 2 (e.g., 128) as follows (please make sure it's a power of 2):
```
bash /home/shicao/eval/scripts/run-exp-n1-segfault-no-fusion.sh 128
```

If you want a multi-node reproducer (2 nodes), you can try the following command while you can also increase the first argument to a bigger power of 2 to make SegFault more deterministic:
```
bash /home/shicao/eval/scripts/run-exp-n1-segfault-no-fusion.sh 64
```

**Comment by MadFunMaker** [comment]
Good news is that I think GDB backtrace now seems to be more fine-grained (thus more informative hopefully) to me! Here are different GDB backtraces I faced so far while I was trying to minimize a reproducer in case these help us narrow down the issue.

## Reproducer
```
bash run-exp-n1-segfault-no-fusion.sh 64
```

## Backtrace 1

```
#0  0x00007fd65e1ff23f in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x55c4c3aad920, rem=rem@entry=0x55c4c3aad920) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007fd65e204ec7 in __GI___nanosleep (requested_time=requested_time@entry=0x55c4c3aad920, remaining=remaining@entry=0x55c4c3aad920) at nanosleep.c:27
#2  0x00007fd65e204dfe in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x00007fd65f63cd45 in Realm::realm_freeze (signal=11) at /home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc:183
#4  <signal handler called>
#5  0x00007fd663ca5941 in Legion::Internal::SingleTask::launch_task (this=0x7fc41dc36690, inline_task=false) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:4719
#6  0x00007fd663cb101e in Legion::Internal::PointTask::complete_replay (this=0x7fc41dc36690, instance_ready_event=..., completion_postcondition=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:7560
#7  0x00007fd663d2b517 in Legion::Internal::CompleteReplay::execute (this=0x7fc42c41e7d0, events=std::vector of length 41500, capacity 65536 = {...}, Python Exception <class 'gdb.error'> Attempt to extract a component of a value that is not a struct/class/union.:
user_events=std::map with 59 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer':
operations=std::map with 6176 elements,
    recurrent_replay=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:10166
#8  0x00007fd663d0ef8a in Legion::Internal::PhysicalTemplate::execute_slice (this=0x7fc4103ebcc0, slice_idx=1, recurrent_replay=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:4729
#9  0x00007fd663d1de6c in Legion::Internal::PhysicalTemplate::handle_replay_slice (args=0x7fc41cc46890) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:7595
#10 0x00007fd663f5bfa9 in Legion::Internal::Runtime::legion_runtime_task (args=0x7fc41cc46890, arglen=20, userdata=0x55c4c3a43610, userlen=8, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.cc:32456
#11 0x00007fd65f613e78 in Realm::LocalTaskProcessor::execute_task (this=0x55c4c0551320, func_id=4, task_args=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/proc_impl.cc:1129
#12 0x00007fd65f68c3ed in Realm::Task::execute_on_processor (this=0x7fc41cc46710, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:302
#13 0x00007fd65f6915d8 in Realm::UserThreadTaskScheduler::execute_task (this=0x55c4c0551630, task=0x7fc41cc46710) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:1632
#14 0x00007fd65f68f380 in Realm::ThreadedTaskScheduler::scheduler_loop (this=0x55c4c0551630) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:1105
#15 0x00007fd65f697e4e in Realm::Thread::thread_entry_wrapper<Realm::ThreadedTaskScheduler, &Realm::ThreadedTaskScheduler::scheduler_loop> (obj=0x55c4c0551630) at /home/shicao/ff-gpp/deps/legion/runtime/realm/threads.inl:97
#16 0x00007fd65f6a6c56 in Realm::UserThread::uthread_entry () at /home/shicao/ff-gpp/deps/legion/runtime/realm/threads.cc:1355
#17 0x00007fd65e17d4e0 in ?? () at ../sysdeps/unix/sysv/linux/x86_64/__start_context.S:91 from /lib/x86_64-linux-gnu/libc.so.6
#18 0x0000000000000000 in ?? ()
```

## Backtrace 2

```
#0  0x00007fd5e1e7a23f in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x55fd53e3b920, rem=rem@entry=0x55fd53e3b920) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007fd5e1e7fec7 in __GI___nanosleep (requested_time=requested_time@entry=0x55fd53e3b920, remaining=remaining@entry=0x55fd53e3b920) at nanosleep.c:27
#2  0x00007fd5e1e7fdfe in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x00007fd5e32b7d45 in Realm::realm_freeze (signal=6) at /home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc:183
#4  <signal handler called>
#5  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#6  0x00007fd5e1dbf859 in __GI_abort () at abort.c:79
#7  0x00007fd5e1dbf729 in __assert_fail_base (fmt=0x7fd5e1f55588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x7fd5e45532a0 "0 && \"invalid reservation handle\"",
    file=0x7fd5e4551c78 "/home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc", line=2645, function=<optimized out>) at assert.c:92
#8  0x00007fd5e1dd0fd6 in __GI___assert_fail (assertion=0x7fd5e45532a0 "0 && \"invalid reservation handle\"", file=0x7fd5e4551c78 "/home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc", line=2645,
    function=0x7fd5e4553210 "Realm::ReservationImpl* Realm::RuntimeImpl::get_lock_impl(Realm::ID)") at assert.c:101
#9  0x00007fd5e32c3825 in Realm::RuntimeImpl::get_lock_impl (this=0x55fd2d9580e0, id=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc:2645
#10 0x00007fd5e32acfc5 in Realm::Reservation::release (this=0x7fc39dfedfc8, wait_on=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/rsrv_impl.cc:220
#11 0x00007fd5e777a8f3 in Legion::Internal::Runtime::release_reservation (r=..., precondition=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.h:5731
#12 0x00007fd5e7920949 in Legion::Internal::SingleTask::launch_task (this=0x7fc3ae5e1140, inline_task=false) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:4719
#13 0x00007fd5e792c01e in Legion::Internal::PointTask::complete_replay (this=0x7fc3ae5e1140, instance_ready_event=..., completion_postcondition=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:7560
#14 0x00007fd5e79a6517 in Legion::Internal::CompleteReplay::execute (this=0x7fc3a882efb0, events=std::vector of length 41500, capacity 65536 = {...}, Python Exception <class 'gdb.error'> Attempt to extract a component of a value that is not a struct/class/union.:
user_events=std::map with 59 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer':
operations=std::map with 6176 elements,
    recurrent_replay=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:10166
#15 0x00007fd5e7989f8a in Legion::Internal::PhysicalTemplate::execute_slice (this=0x7fc3b8580d20, slice_idx=0, recurrent_replay=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:4729
#16 0x00007fd5e7998e6c in Legion::Internal::PhysicalTemplate::handle_replay_slice (args=0x7fc2e80028c0) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:7595
#17 0x00007fd5e7bd6fa9 in Legion::Internal::Runtime::legion_runtime_task (args=0x7fc2e80028c0, arglen=20, userdata=0x55fd2dd05410, userlen=8, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.cc:32456
#18 0x00007fd5e328ee78 in Realm::LocalTaskProcessor::execute_task (this=0x55fd50932f10, func_id=4, task_args=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/proc_impl.cc:1129
#19 0x00007fd5e33073ed in Realm::Task::execute_on_processor (this=0x7fc2e8002740, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:302
#20 0x00007fd5e330c5d8 in Realm::UserThreadTaskScheduler::execute_task (this=0x55fd303df490, task=0x7fc2e8002740) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:1632
#21 0x00007fd5e330a380 in Realm::ThreadedTaskScheduler::scheduler_loop (this=0x55fd303df490) at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:1105
#22 0x00007fd5e3312e4e in Realm::Thread::thread_entry_wrapper<Realm::ThreadedTaskScheduler, &Realm::ThreadedTaskScheduler::scheduler_loop> (obj=0x55fd303df490) at /home/shicao/ff-gpp/deps/legion/runtime/realm/threads.inl:97
#23 0x00007fd5e3321c56 in Realm::UserThread::uthread_entry () at /home/shicao/ff-gpp/deps/legion/runtime/realm/threads.cc:1355
#24 0x00007fd5e1df84e0 in ?? () at ../sysdeps/unix/sysv/linux/x86_64/__start_context.S:91 from /lib/x86_64-linux-gnu/libc.so.6
#25 0x0000000000000000 in ?? ()
```

## Backtrace 3

```
#0  0x00007f303b05523f in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0,
    req=req@entry=0x55c361c90920, rem=rem@entry=0x55c361c90920) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007f303b05aec7 in __GI___nanosleep (requested_time=requested_time@entry=0x55c361c90920,
    remaining=remaining@entry=0x55c361c90920) at nanosleep.c:27
#2  0x00007f303b05adfe in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x00007f303c492d45 in Realm::realm_freeze (signal=11)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/runtime_impl.cc:183
#4  <signal handler called>
#5  std::__atomic_base<unsigned int>::fetch_or (__m=std::memory_order_acq_rel, __i=1, this=0x18)
    at /usr/include/c++/9/bits/atomic_base.h:571
#6  Realm::atomic<unsigned int>::fetch_or_acqrel (this=0x18, to_or=1)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/atomics.inl:258
#7  0x00007f303be2219f in Realm::UnfairMutex::lock (this=0x18)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/mutex.inl:171
#8  0x00007f303be24d07 in Realm::AutoLock<Realm::UnfairMutex>::AutoLock (this=0x7f1e0b4315a0, mutex=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/mutex.inl:329
#9  0x00007f303c488eca in Realm::ReservationImpl::acquire (this=0x0, new_mode=0, exclusive=true,
    acquire_type=Realm::ReservationImpl::ACQUIRE_BLOCKING, after_lock=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/rsrv_impl.cc:375
#10 0x00007f303c48a519 in Realm::FastRsrvState::request_base_rsrv (this=0x7f1d20708618, frsv=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/rsrv_impl.cc:772
#11 0x00007f303c48ab42 in Realm::FastReservation::wrlock_slow (this=0x7f1d20708618, mode=Realm::FastReservation::SPIN)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/rsrv_impl.cc:968
#12 0x00007f304056acaf in Realm::FastReservation::wrlock (this=0x7f1d20708618, mode=Realm::FastReservation::SPIN)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/reservation.inl:70
#13 0x00007f304056b024 in Legion::Internal::LocalLock::wrlock (this=0x7f1d20708618)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_types.h:2725
#14 0x00007f304056b1a1 in Legion::Internal::AutoLock::AutoLock (this=0x7f1e0b431bb0, r=..., mode=0, excl=true)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_types.h:2757
#15 0x00007f3040d9f4b3 in Legion::Internal::Runtime::increment_total_outstanding_tasks (this=0x7f1d207084f8, tid=48,
    meta=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.cc:27611
#16 0x00007f3040b22995 in Legion::Internal::Runtime::issue_runtime_meta_task<Legion::Internal::SingleTask::MisspeculationTaskArgs> (this=0x7f1d207084f8, args=..., priority=Legion::Internal::LG_LATENCY_WORK_PRIORITY, precondition=..., target=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.h:5058
#17 0x00007f3040afbb10 in Legion::Internal::SingleTask::launch_task (this=0x7f1d20874010, inline_task=false)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:4754
#18 0x00007f3040b0701e in Legion::Internal::PointTask::complete_replay (this=0x7f1d20874010, instance_ready_event=...,
    completion_postcondition=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_tasks.cc:7560
#19 0x00007f3040b81517 in Legion::Internal::CompleteReplay::execute (this=0x7f1e05478230,
    events=std::vector of length 83004, capacity 131072 = {...}, Python Exception <class 'gdb.error'> Attempt to extract a component of a value that is not a struct/class/union.:
user_events=std::map with 123 elements, Python Exception <class 'AttributeError'> 'NoneType' object has no attribute 'pointer':

    operations=std::map with 12352 elements, recurrent_replay=true)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:10166
#20 0x00007f3040b64f8a in Legion::Internal::PhysicalTemplate::execute_slice (this=0x7f1e1091f2e0, slice_idx=0,
    recurrent_replay=true) at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:4729
#21 0x00007f3040b73e6c in Legion::Internal::PhysicalTemplate::handle_replay_slice (args=0x7f1d4d7cd4a0)
    at /home/shicao/ff-gpp/deps/legion/runtime/legion/legion_trace.cc:7595
#22 0x00007f3040db1fa9 in Legion::Internal::Runtime::legion_runtime_task (args=0x7f1d4d7cd4a0, arglen=20,
    userdata=0x55c361c47b00, userlen=8, p=...) at /home/shicao/ff-gpp/deps/legion/runtime/legion/runtime.cc:32456
#23 0x00007f303c469e78 in Realm::LocalTaskProcessor::execute_task (this=0x55c35e7550e0, func_id=4, task_args=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/proc_impl.cc:1129
#24 0x00007f303c4e23ed in Realm::Task::execute_on_processor (this=0x7f1d4d7cd320, p=...)
    at /home/shicao/ff-gpp/deps/legion/runtime/realm/tasks.cc:302
```

**Comment by lightsighter** [comment]
Those backtraces are actually an unrelated bug that has already been fixed. Can you update to the most recent control replication branch?

> the minimal one only requires a single node now with much less number of tasks

Now that you can reproduce it on a single node, can you run that configuration with valgrind and see what it reports? Valgrind will be able to tell us as soon as anything does an illegal memory write. Ignore any warning that you get from valgrind, just pay attention to the actual errors.

> From my experience, you can increase the chance of SegFault by increasing the first argument (loosely, the number of tasks) to a bigger power of 2 (e.g., 128) as follows (please make sure it's a power of 2):

That would be consistent with an issue in the transitive reduction. The bigger the graph of instructions to optimize the more it will walk off the end of some data structure and corrupt the heap.

**Comment by MadFunMaker** [comment]
> Those backtraces are actually an unrelated bug that has already been fixed. Can you update to the most recent control replication branch?

I still face the same error messages after I updated it to the most recent control replication branch. 

> Now that you can reproduce it on a single node, can you run that configuration with valgrind and see what it reports? Valgrind will be able to tell us as soon as anything does an illegal memory write. Ignore any warning that you get from valgrind, just pay attention to the actual errors.

What I reproduce with a single node is actually an unrelated bug that you said has already been fixed. But, as I just said, this bug still occurs. What can we do to fix this unrelated bug? After we fix this, I can retry to reproduce our original SegFault in a single node in DEBUG mode without any optimization and run Valgrind as you suggested.

**Comment by lightsighter** [comment]
Just run with valgrind first on a single node and see what it says.
